openage/convert/smx.pyx

```diff
@@ -0,0 +1,1145 @@
+        cdef uint8_t pixel_mask_even_2 = 0b11110000
+        cdef uint8_t pixel_mask_even_3 = 0b00111111
+
+        # work through commands till end of row.
+        while not eor:
+            if row_data.size() > expected_size:
+                raise Exception(
+                    "Only %d pixels should be drawn in row %d "
+                    "with frame type %d, but we have %d "
+                    "already!" % (
+                        expected_size, rowid,
+                        self.info.frame_type,
+                        row_data.size()
+                    )
+                )
+
+            # fetch drawing instruction
+            cmd = self.get_byte_at(dpos_cmd)
+
+            # Last 2 bits store command type
+            lower_crumb = 0b00000011 & cmd
+
+            if lower_crumb == 0b00000011:
+                # eor (end of row) command, this row is finished now.
+                eor = True
+                dpos_cmd += 1
+
+                continue
+
+            elif lower_crumb == 0b00000000:
+                # skip command
+                # draw 'count' transparent pixels
+                # count = (cmd >> 2) + 1
+
+                pixel_count = (cmd >> 2) + 1
+
+                for _ in range(pixel_count):
+                    row_data.push_back(pixel(color_transparent, 0, 0, 0, 0))
+
+            elif lower_crumb == 0b00000001:
+                # color_list command
+                # draw the following 'count' pixels
+                # pixels are stored in 5 byte chunks
+                # even pixel indices have their info stored
+                # in byte[0] - byte[3]. odd pixel indices have
+                # their info stored in byte[1] - byte[4].
+                # count = (cmd >> 2) + 1
+
+                pixel_count = (cmd >> 2) + 1
+
+                for _ in range(pixel_count):
+                    # Start fetching pixel data
+                    if odd:
+                        # Odd indices require manual extraction of each of the 4 values
+
+                        # Palette index. Essentially a rotation of (byte[1]byte[2])
+                        # by 6 to the left, then masking with 0x00FF.
+                        pixel_data_odd_0 = self.get_byte_at(dpos_color + 1)
+                        pixel_data_odd_1 = self.get_byte_at(dpos_color + 2)
+                        pixel_data.push_back((pixel_data_odd_0 >> 2) | (pixel_data_odd_1 << 6))
+
+                        # Palette section. Described in byte[2] in bits 4-5.
+                        pixel_data.push_back((pixel_data_odd_1 >> 2) & 0x03)
+
+                        # Damage mask 1. Essentially a rotation of (byte[3]byte[4])
+                        # by 6 to the left, then masking with 0x00FF.
+                        pixel_data_odd_2 = self.get_byte_at(dpos_color + 3)
+                        pixel_data_odd_3 = self.get_byte_at(dpos_color + 4)
+                        pixel_data.push_back((pixel_data_odd_0 >> 2) | (pixel_data_odd_1 << 6))
```

Comment:
hmm, should this be pixel_data_odd_2/3?

Reply:
Yes, thank you! There was also an issue with the masking there which is now fixed.